packages/fxa-payments-server/src/lib/validator.test.tsx

```diff
@@ -0,0 +1,171 @@
+  });
+});
+
+it('supports updating a field', () => {
+  const { results } = runAgainstValidator(
+    v => v.registerField({ name: 'foo', fieldType: 'input', required: true }),
+    v => v.updateField({ name: 'foo', value: 'bar', valid: false, error: 'meep' }),
+    v => v.getFieldState('foo'),
+  );
+  expect(results.pop()).toEqual({
+    fieldType: 'input',
+    required: true,
+    value: 'bar',
+    valid: false,
+    error: 'meep'
+  });
+});
+
+it('supports checking if fields are valid', () => {
+  const { results } = runAgainstValidator(
+    v => v.registerField({ name: 'foo', fieldType: 'input', required: true }),
+    v => v.registerField({ name: 'baz', fieldType: 'input', required: true }),
+    v => v.updateField({ name: 'foo', value: 'bar', valid: true }),
+    v => v.updateField({ name: 'baz', value: 'quux', valid: false, error: 'meep' }),
+    v => v.isInvalid('baz'),
+    v => v.getError('baz'),
+    v => v.allValid(),
+    v => v.updateField({ name: 'baz', value: 'xyzzy', valid: true }),
+    v => v.allValid(),
+  );
+  expect(results).toEqual([
+    undefined,
+    undefined,
+    undefined,
+    undefined,
+    true,
+    'meep',
+    false,
+    undefined,
+    true
+  ]);
+});
+
+it('supports getting field values', () => {
+  const { results } = runAgainstValidator(
+    v => v.registerField({ name: 'foo', fieldType: 'input', required: true }),
+    v => v.registerField({ name: 'baz', fieldType: 'input', required: true }),
+    v => v.registerField({ name: 'hello', fieldType: 'input', required: true }),
+    v => v.registerField({ name: 'nothere', fieldType: 'input', required: true }),
+    v => v.updateField({ name: 'foo', value: 'bar', valid: true }),
+    v => v.updateField({ name: 'baz', value: 'xyzzy', valid: true }),
+    v => v.updateField({ name: 'hello', value: 'world', valid: true }),
+    v => v.getValue('foo'),
+    v => v.getValue('baz'),
+    v => v.getValue('hello'),
+    v => v.getValue('nothere'),
+    v => v.getValue('nothere', 'honk'),
+    v => v.getValues(),
+  );
+  expect(results).toEqual([
+    undefined,
+    undefined,
+    undefined,
+    undefined,
+    undefined,
+    undefined,
+    undefined,
+    'bar',
+    'xyzzy',
+    'world',
+    undefined,
+    'honk',
+    {
+      foo: 'bar',
+      baz: 'xyzzy',
+      hello: 'world',
+      nothere: null,
+    }
+  ]);
+});
+
+it('supports set, get, reset of a global error', () => {
+  const { results } = runAgainstValidator(
+    v => v.getGlobalError(),
+    v => v.setGlobalError('everything is bad'),
+    v => v.getGlobalError(),
+    v => v.resetGlobalError(),
+    v => v.getGlobalError(),
+  );
+  expect(results).toEqual([
+    null,
+    undefined,
+    'everything is bad',
+    undefined,
+    null,
+  ]);
+});
+
+// runAgainstValidator is kind of a funky hack, but this seems like the least
+// painful way to exercise a useReducer() hook somewhat realistically.
+const runAgainstValidator = (...fns: Array<(validator: Validator) => any>) => {
```

Comment:
The way these tests work is a little funky - since React Testing Library doesn't lend direct access to objects & properties under the hood, I hack around that by dumping results & validator state as JSON into an element's text content. And I trigger test functions by way of clicking buttons. Kind of clunky, but it does exercise the whole assembly of fields & validator state in React rendering.